### PR TITLE
keywords: add list protocols parameter and output

### PR DESCRIPTION
**Pull request by norg**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:

https://redmine.openinfosecfoundation.org/issues/635

Describe changes:
- this introduces "--list-protocols" to list additional keywords specific to protocols
- Follow up of #3938 

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output (if applicable):

- PR norg-pcap: https://buildbot.openinfosecfoundation.org/builders/norg-pcap/builds/69
- PR norg: https://buildbot.openinfosecfoundation.org/builders/norg/builds/71

### CODE REVIEW

**Comment by catenacyber** [comment]
Thanks @norg 
This looks good.
How did you build this list ?
I am wondering how we can add a test so that adding a protocol requires changing  `ListSupportedDetectProtocols`

**Comment by norg** [comment]
I just grepped it from the source code, nothing fancy

**Comment by catenacyber** [comment]
Cool. Do you have the syntax of your `grep` ?

**Comment by norg** [comment]
Was nothing fancy, just using the ones I knew were missing and looking if they are somewhere else besides **src/detect-engine-proto.c**

**Comment by catenacyber** [comment]
Should we add a test diffing `grep strcasecmp src/detect-engine-proto.c | cut -d'"' -f2` and `suricata list-detect-protocols` ?

**Comment by norg** [comment]
Could be a start for sure

**Comment by jlucovsky** [comment]
> Should we add a test diffing `grep strcasecmp src/detect-engine-proto.c | cut -d'"' -f2` and `suricata list-detect-protocols` ?

I like the idea of this. I'd prefer the implementation be driven from a table used by `DetectProtoParse` to avoid drift between `--list-protocol` output and the protocols handled in `DetectProtoParse`.

Additionally, it's confusing to see `tcp-pkt` and `tcp-stream` (and similar) included in a `protocol` list. These aren't protocols.

File: src/detect-engine-proto.c [hunk 1]
```diff
@@ -143,6 +143,29 @@ int DetectProtoContainsProto(const DetectProto *dp, int proto)
     return 0;
 }
 
+/* List supported detect protocol keywords */
+
+void ListSupportedDetectProtocols(void)
+{
+    SCEnter();
+    printf("=========Supported Detect Protocols=========\n");
+	printf("- tcp\n");
```
Comment (victorjulien) [inline src/detect-engine-proto.c]: indent looks odd here

**Comment by victorjulien** [comment]
These are valid values for the signature's protocol field. It think its fine to list tcp-pkt/tcp-stream here.

File: src/detect-engine-proto.c [hunk 2]
```diff
@@ -143,6 +143,29 @@ int DetectProtoContainsProto(const DetectProto *dp, int proto)
...
+	printf("- icmpv6\n");
+	printf("- icmp\n");
+	printf("- sctp\n");
+	printf("- ipv4\n");
+	printf("- ip4\n");
+	printf("- ipv6\n");
+	printf("- ip6\n");
+	printf("- ip\n");
+	printf("- pkthdr\n");
+    SCReturn;
```
Comment (victorjulien) [inline src/detect-engine-proto.c]: Should this also list the valid app protos?

File: src/detect-engine-proto.c (hunk 1 above)
Comment (norg) [reply src/detect-engine-proto.c]: Hmm need to double check in my vim as it looks ok there, but should be easy to fix.

File: src/detect-engine-proto.c (hunk 2 above)
Comment (norg) [reply src/detect-engine-proto.c]: within this list but with another headline?

**Comment by catenacyber** [comment]
What is the status on this ?
An updated PR is expected ?

**Pull request closed without merging**